src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -149,6 +141,206 @@ public interface IFileWatchingToken
         {
         }
 
+        private readonly struct WatcherOperation
+        {
+            private readonly Kind _kind;
+
+            private readonly string _directory;
+            private readonly string? _filter;
+            private readonly _VSFILECHANGEFLAGS _fileChangeFlags;
+            private readonly IVsFreeThreadedFileChangeEvents2 _sink;
+            private readonly List<uint> _cookies;
+
+            private readonly IFileWatchingToken _token;
+            private readonly IEnumerable<IFileWatchingToken> _tokens;
```

Comment:
i'm curious why a readonly, flattened, struct vs a base type with derived types.  i only ask because looking at the code below, it looks like a lot of work knowing which fields correspond to which kinds.  

If we do the flattened form, it would be good to potentially doc each of these as to which kinds they apply to.

Reply:
> readonly, flattened, struct

I wanted to minimize the allocations considering `TryCombineWith` is called on every item in the batch.

Docs will be added for these once they are all complete. 👍

Reply:
➡️ `TryCombineWith` is now implemented, and the fields are documented.